### PR TITLE
Manipulator

### PR DESCRIPTION
**Pull request by MatthewChoulas**
re opening draft for this

### CODE REVIEW

**Review by eckshn — COMMENTED** [review]
please add arm motor and change as necessary to implement

File: src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeConeCommand.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package com.team4099.robot2022.commands.intake
+
+import com.team4099.robot2023.config.constants.ManipulatorConstants
+import com.team4099.robot2023.subsystems.manipulator.Manipulator
+import edu.wpi.first.wpilibj2.command.CommandBase
+import org.littletonrobotics.junction.Logger
+
+class IntakeConeCommand(val manipulator: Manipulator) : CommandBase() {
```
Comment (eckshn) [inline src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeConeCommand.kt:8]: for each command: add armState

File: src/main/kotlin/com/team4099/robot2023/config/constants/ManipulatorConstants.kt [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+  // TODO(figure out what voltage compensation should be)
+  val ARM_VOLTAGE_COMPENSATION = 10.0.volts
+  val INTAKE_VOLTAGE_COMPENSATION = 10.0.volts
+
+  val ARM_SPOOL_RADIUS = 1.0.inches
+  // TODO(What is the inertial value)
+  val MOMENT_INERTIA = 0.0045
+
+  enum class RollerState(val speed: Double) {
```
Comment (eckshn) [inline src/main/kotlin/com/team4099/robot2023/config/constants/ManipulatorConstants.kt:49]: create an enum of ArmStates.

Comment (eckshn) [reply src/main/kotlin/com/team4099/robot2023/config/constants/ManipulatorConstants.kt:49]: ask design to figure out how far arm needs to extend for everything

File: src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIO.kt [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        armTempCelcius = it.celsius
+      }
+    }
+  }
+
+  fun updateInputs(inputs: ManipulatorIOInputs) {}
+
+  fun setRollerPower(percentOutput: Double) {}
+
+  fun setOpenLoop(percentOutput: Double) {}
```
Comment (eckshn) [inline src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIO.kt:96]: fill out the functions

File: src/main/kotlin/com/team4099/robot2023/config/constants/Constants.kt [hunk 4]
```diff
@@ -62,6 +62,10 @@ object Constants {
     const val PIGEON_2_ID = 1337
   }
 
+  object Manipulator {
+    const val INTAKE_MOTOR_ID = 0
```
[the remark this answers is not in the document]
Comment (eckshn) [reply src/main/kotlin/com/team4099/robot2023/config/constants/Constants.kt:66]: in the google drive in 2023, there is a sheet called JVN that has a sheet that has all of our can IDs

File: build.gradle [hunk 5]
```diff
@@ -9,8 +9,8 @@ plugins {
 	id "org.jetbrains.kotlinx.kover" version "0.4.2"
 }
 
-sourceCompatibility = JavaVersion.VERSION_11
-targetCompatibility = JavaVersion.VERSION_11
+sourceCompatibility = JavaVersion.VERSION_17
```
Comment (plusparth) [inline build.gradle:12]: These types of changes should go into `main` and not be in this PR. Because you have done these in separate commits anyway, you can just `cherry-pick` the commits into a new branch, make a separate PR, and merge that before this thing is ready

Comment (plusparth) [reply build.gradle:12]: let me know if you have questions about this -- it should be easy and fast once you understand what you have to do

File: src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeConeCommand.kt (hunk 1 above)
Comment (plusparth) [reply src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeConeCommand.kt:8]: This will later have to use the elevator too right?

File: src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeCubeCommand.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package com.team4099.robot2022.commands.intake
+
+import com.team4099.robot2023.config.constants.ManipulatorConstants
+import com.team4099.robot2023.subsystems.manipulator.Manipulator
+import edu.wpi.first.wpilibj2.command.CommandBase
+import org.littletonrobotics.junction.Logger
+
+class IntakeCubeCommand(val manipulator: Manipulator) : CommandBase() {
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeCubeCommand.kt:8]: What you set the states to will depend heavily on where you are intaking from -- human player station cube intake will be a different manipulator state than ground intake, bc the manipulator position will likely be different

File: src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeCubeCommand.kt [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  override fun execute() {
+    Logger.getInstance().recordOutput("ActiveCommands/IntakeCobeCommand", true)
+  }
+
+  override fun end(interrupted: Boolean) {
+    manipulator.rollerState = ManipulatorConstants.RollerState.IDLE
+  }
+
+  override fun isFinished(): Boolean {
+    return false
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeCubeCommand.kt:27]: Is this what we want here? We should be able to have logic here (both for this and the cone intake command) that can detect when we have grabbed a game piece, and end the command there. This will be very useful for chaining this command with other actions for driver assist features

File: src/main/kotlin/com/team4099/robot2023/commands/manipulator/ManipulatorIdleCommand.kt [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  init {
+    addRequirements(manipulator)
+  }
+
+  override fun initialize() {
+    manipulator.rollerState = ManipulatorConstants.RollerState.IDLE
+  }
+
+  override fun execute() {
+    Logger.getInstance().recordOutput("ActiveCommands/IntakeIdleCommand", true)
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/commands/manipulator/ManipulatorIdleCommand.kt:18]: This isn't an intake idle command (rename the field)

File: src/main/kotlin/com/team4099/robot2023/commands/manipulator/OuttakeConeCommand.kt [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    addRequirements(manipulator)
+  }
+
+  override fun initialize() {
+
+    manipulator.rollerState = ManipulatorConstants.RollerState.CONE_OUT
+  }
+
+  override fun execute() {
+    Logger.getInstance().recordOutput("ActiveCommands/ReverseConeCommand", true)
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/commands/manipulator/OuttakeConeCommand.kt:19]: Make sure the name of what you log matches the name of the command, otherwise it will make debugging unnecessarily difficult

File: src/main/kotlin/com/team4099/robot2023/commands/manipulator/OutttakeCubeCommand.kt [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    addRequirements(manipulator)
+  }
+
+  override fun initialize() {
+
+    manipulator.rollerState = ManipulatorConstants.RollerState.CUBE_OUT
+  }
+
+  override fun execute() {
+    Logger.getInstance().recordOutput("ActiveCommands/ReverseCobeCommand", true)
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/commands/manipulator/OutttakeCubeCommand.kt:19]: Same as above

File: src/main/kotlin/com/team4099/robot2023/config/constants/FieldConstants.kt [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+
+/**
+ * Contains various field dimensions and useful reference points. Dimensions are in meters, and sets
+ * of corners start in the lower left moving clockwise.
+ *
+ * All translations and poses are stored with the origin at the rightmost point on the BLUE ALLIANCE
+ * wall. Use the [.allianceFlip] and [.allianceFlip] methods to flip these values based on the
+ * current alliance color.
+ */
+object FieldConstants {
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/config/constants/FieldConstants.kt:25]: Same as above -- this should be in a separate PR

File: src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/Manipulator.kt [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+import org.team4099.lib.units.base.inSeconds
+import org.team4099.lib.units.derived.inVoltsPerMeter
+import org.team4099.lib.units.derived.inVoltsPerMeterPerSecond
+import org.team4099.lib.units.derived.inVoltsPerMeterSeconds
+import org.team4099.lib.units.derived.perMeter
+import org.team4099.lib.units.derived.perMeterPerSecond
+import org.team4099.lib.units.derived.perMeterSeconds
+import org.team4099.lib.units.derived.volts
+
+class Manipulator(val io: ManipulatorIO) : SubsystemBase() {
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/Manipulator.kt:18]: Is there a way for this subsystem to be aware of which game piece it has? (hint: there is). I think it is harder to know if it currently is holding something, but that should be doable too.

File: src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIONeo.kt [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+  override fun zeroEncoder() {
+    armSparkMax.encoder.position = 0.0
+  }
+
+  override fun configPID(
+    kP: ProportionalGain<Meter, Volt>,
+    kI: IntegralGain<Meter, Volt>,
+    kD: DerivativeGain<Meter, Volt>
+  ) {
+    armPIDController.p = kP.inVoltsPerMeter
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIONeo.kt:111]: This is not correct -- the Spark MAX PID controller does not take PID values in these units. You need to use your linear mechanism sensor to convert the PID units to native units

File: src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIONeo.kt [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    // percentOutput * statorCurrent
+    inputs.armSupplyCurrent = inputs.armStatorCurrent * armSparkMax.appliedOutput
+    inputs.armTempCelcius = armSparkMax.motorTemperature.celsius
+  }
+  override fun setOpenLoop(percentOutput: Double) {
+    armSparkMax.set(percentOutput)
+  }
+
+  override fun setPosition(position: Length, feedforward: ElectricalPotential) {
+    armPIDController.setFF(feedforward.inVolts)
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIONeo.kt:99]: This will not set the arm to a position (I assume you know this and it is just incomplete lol)

**Review by plusparth — CHANGES_REQUESTED** [review]
I skimmed so could have missed stuff, but looks good overall 👍  We should definitely clean up some of the PRs by moving unrelated stuff out of them, though -- it makes reviewing way easier, which means it reduces the likelihood of bugs

File: src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeConeCommand.kt (hunk 1 above)
Comment (sswadkar) [reply src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeConeCommand.kt:8]: if we're keeping subsystems separate then it makes sense to schedule this command after elevator has gone to SHELF_HEIGHT or smthing. unless we want to add a precondition to make sure the elevator is at the right height before running this command? but i don't see any particular use case for that

File: src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIONeo.kt (hunk 13 above)
Comment (sswadkar) [reply src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIONeo.kt:111]: an example can be found in drivetrain code in SwerveIOFalcon init

File: src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIOSim.kt [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+object ManipulatorIOSim : ManipulatorIO {
+
+  val manipulatorMotor: FlywheelSim =
+    FlywheelSim(
+      DCMotor.getNEO(1), ManipulatorConstants.GEAR_RATIO, ManipulatorConstants.MOMENT_INERTIA
+    )
+  val mechanism_widget: Mechanism2d = Mechanism2d(700.0, 700.0, Color8Bit(Color.kBlack))
+  val mechanism_root: MechanismRoot2d = mechanism_widget.getRoot("pivot", 350.0, 0.0)
+  val mechanism_arm: MechanismLigament2d =
+    mechanism_root.append(MechanismLigament2d("arm", 700.0, 0.0, 8.0, Color8Bit(Color.kWhite)))
```
Comment (sswadkar) [inline src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIOSim.kt:28]: do these numbers mean anything? can we unitify them or nah

**Review by sswadkar — CHANGES_REQUESTED** [review]
yall are incredible. amazing job j left a few comments for general improvement. also please go through and add doc strings and other comments wherever you can :).

File: src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIONeo.kt [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+    intakeSparkMax.setIdleMode(CANSparkMax.IdleMode.kCoast)
+    intakeSparkMax.inverted = ManipulatorConstants.INTAKE_MOTOR_INVERTED
+    intakeSparkMax.burnFlash()
+    intakeSparkMax.openLoopRampRate = ManipulatorConstants.INTAKE_RAMP_RATE
+
+    armSparkMax.enableVoltageCompensation(ManipulatorConstants.ARM_VOLTAGE_COMPENSATION.inVolts)
+    armSparkMax.setSmartCurrentLimit(ManipulatorConstants.ARM_SUPPLY_CURRENT_LIMIT)
+    armSparkMax.inverted = ManipulatorConstants.ARM_MOTOR_INVERTED
+    armSparkMax.burnFlash()
+    armSparkMax.openLoopRampRate = ManipulatorConstants.ARM_RAMP_RATE
```
[the remark this answers is not in the document]
Comment (sswadkar) [reply src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIONeo.kt:64]: also worth adding a brake mode function to help config brake mode from robot container (like how we did w drive motors in code last year)

File: build.gradle (hunk 5 above)
Comment (MatthewChoulas) [reply build.gradle:12]: this was rebased from main

File: src/main/kotlin/com/team4099/robot2023/config/constants/FieldConstants.kt (hunk 11 above)
Comment (MatthewChoulas) [reply src/main/kotlin/com/team4099/robot2023/config/constants/FieldConstants.kt:25]: i think this was also rebased at some point, idk why it shows up as an addition

File: src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIOSim.kt (hunk 15 above)
Comment (MatthewChoulas) [reply src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIOSim.kt:28]: pls nah

File: build.gradle (hunk 5 above)
Comment (plusparth) [reply build.gradle:12]: If you rebase onto a branch that has this commit, it shouldn't show up in the diff here -- so we're doing something wrong here

File: src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIOSim.kt (hunk 15 above)
Comment (plusparth) [reply src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIOSim.kt:33]: why are these underscores?

Comment (plusparth) [reply src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/ManipulatorIOSim.kt:33]: also, is there a way to make the "root" be reused from elevator after this merges, so you can see the arm move with the elevator?

File: src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeConeCommand.kt (hunk 1 above)
Comment (plusparth) [reply src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeConeCommand.kt:7]: yeah, that's fine -- I guess the point really was that there will have to be command groups later that chain these in the right order

File: src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/Manipulator.kt [hunk 17]
```diff
@@ -50,27 +50,20 @@ class Manipulator(val io: ManipulatorIO) : SubsystemBase() {
...
-  val intakingGamePiece: Boolean
+  val hasCube: Boolean
     get() {
-      return inputs.rollerStatorCurrent >= ManipulatorConstants.INTAKE_CURRENT_THRESHOLD &&
-        (
-          rollerState == ManipulatorConstants.RollerStates.CUBE_IN ||
-            rollerState == ManipulatorConstants.RollerStates.CONE_IN
-          ) &&
+      return inputs.rollerStatorCurrent >= ManipulatorConstants.CUBE_CURRENT_THRESHOLD &&
+        rollerState == ManipulatorConstants.RollerStates.CUBE_IN &&
```
Comment (plusparth) [inline src/main/kotlin/com/team4099/robot2023/subsystems/manipulator/Manipulator.kt:56]: What happens after you release the "intake" button, and the roller state is no longer `CUBE_IN`?

File: src/main/kotlin/com/team4099/robot2023/config/constants/ManipulatorConstants.kt [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+  // TODO(check for accuracy)
+  val ARM_MAX_VELOCITY = 30.inches.perSecond
+  val ARM_MAX_ACCELERATION = 15.inches.perSecond.perSecond
+
+  val MOMENT_INERTIA = 0.0000478
+
+  // Tolerance for determining currentRollerState
+  val ROLLER_VOLTAGE_TOLERANCE = 0.4.volts
+
+  enum class RollerStates(val voltage: ElectricalPotential) {
```
Comment (eckshn) [inline src/main/kotlin/com/team4099/robot2023/config/constants/ManipulatorConstants.kt:73]: For all items we need to tune, add a TODO: tune comment so other people will know what could possibly be causing error

File: src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeCubeCommand.kt (hunk 6 above)
Comment (eckshn) [reply src/main/kotlin/com/team4099/robot2023/commands/manipulator/IntakeCubeCommand.kt:8]: building off this, it's worth splitting this command into these different states

File: src/main/kotlin/com/team4099/robot2023/commands/manipulator/ArmCharacterizationCommand.kt [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+  var hasMoved = false
+  var appliedVolts = 0.volts
+  var step = 0.001.volts
+
+  override fun execute() {
+    manipulator.setArmVoltage(appliedVolts)
+
+    if (manipulator.inputs.armVelocity > 0.meters.perSecond) {
+      hasMoved = true
+      println(appliedVolts.inVolts)
```
Comment (Shom770) [inline src/main/kotlin/com/team4099/robot2023/commands/manipulator/ArmCharacterizationCommand.kt:25]: More of a nitpick but shouldn't we be logging this with AdvantageScope rather than printing out specific values? Or is this rather just an artifact of old code
```suggestion
```

File: src/main/kotlin/com/team4099/robot2023/commands/manipulator/ArmCharacterizationCommand.kt [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  var hasMoved = false
+  var appliedVolts = 0.volts
+  var step = 0.001.volts
+
+  override fun execute() {
+    manipulator.setArmVoltage(appliedVolts)
+
+    if (manipulator.inputs.armVelocity < 0.meters.perSecond) {
```
[the remark this answers is not in the document]
Comment (sswadkar) [reply src/main/kotlin/com/team4099/robot2023/commands/manipulator/ArmCharacterizationCommand.kt]: the concern with doing != to is that I highly doubt there's an instant in which you'll pull the `armVelocity` as 0.meters.perSecond. Doing >= means you can find the first instance at which the arm started going. Should add an and condition that checks if it also hasn't moved already tho

[the remark this answers is not in the document]
Comment (eckshn) [reply src/main/kotlin/com/team4099/robot2023/commands/manipulator/ArmCharacterizationCommand.kt]: the AND condition is not necessary as the loop/command would finish before checking again
